client/landing/gutenboarding/path.ts

```diff
@@ -0,0 +1,33 @@
+/**
+ * External dependencies
+ */
+import { generatePath, useRouteMatch } from 'react-router-dom';
+import { getLanguageSlugs } from '../../lib/i18n-utils';
+import { ValuesType } from 'utility-types';
+
+export const Step = {
+	IntentGathering: 'about',
+	DesignSelection: 'design',
+	PageSelection: 'pages',
+	Signup: 'signup',
+	CreateSite: 'create-site',
+} as const;
+
+export const langs: string[] = getLanguageSlugs();
+export const steps = Object.keys( Step ).map( key => Step[ key as keyof typeof Step ] );
+
+export const path = `/:step(${ steps.join( '|' ) })?/:lang(${ langs.join( '|' ) })?`;
+
+export type StepType = ValuesType< typeof Step >;
+
+export function makePath( step: StepType | undefined, lang?: string ) {
+	return generatePath( path, {
+		step,
+		...( lang && langs.includes( lang ) && { lang } ),
+	} );
+}
+
+export function usePath( step: StepType | undefined, lang?: string ) {
+	const match = useRouteMatch< { lang?: string } >( path );
+	return makePath( step, lang || match?.params.lang );
+}
```

Comment:
I suspect we'll hit a case with `usePath` where we want to remove the language. I don't think we can do that with the current implementation and I'm wondering what that would look like.

- `usePath( step, 'fr' )` <- move to step in French
- `usePath( step )` <- move to step with same language
- `usePath( step, XXXXX )` <- How do we say "drop the language"?

Reply:
Would `null` or `false` be good for that?

Reply:
`null` seems like a good way of saying "no language slug"

Reply:
`null` support added here: https://github.com/Automattic/wp-calypso/pull/39312/commits/6929b89db7408edb8441a123c6921675e82a2529